python/cog/server/clients.py

```diff
@@ -0,0 +1,319 @@
+                    throttler.update_last_sent_response_time()
+
+        return sender
+
+    # files
+
+    async def upload_file(
+        self, fh: io.IOBase, *, url: Optional[str], prediction_id: Optional[str]
+    ) -> str:
+        """put file to signed endpoint"""
+        log.debug("upload_file")
+
+        fh.seek(0)
+
+        # try to guess the filename of the given object
+        name = getattr(fh, "name", "file")
+        filename = os.path.basename(name) or "file"
+        assert isinstance(filename, str)
+
+        guess, _ = mimetypes.guess_type(filename)
+        content_type = guess or "application/octet-stream"
+
+        # this code path happens when running outside replicate without upload-url
+        # in that case we need to return data uris
+        if url is None:
+            return file_to_data_uri(fh, content_type)
+        assert url
+
+        # ensure trailing slash
+        url_with_trailing_slash = url if url.endswith("/") else url + "/"
+
+        url = url_with_trailing_slash + filename
+
+        headers = {"Content-Type": content_type}
+        if prediction_id is not None:
+            headers["X-Prediction-ID"] = prediction_id
+
+        # this is a somewhat unfortunate hack, but it works
+        # and is critical for upload training/quantization outputs
+        # if we get multipart uploads working or a separate API route
+        # then we could drop this
+        if url and (".internal" in url or ".local" in url):
+            log.info("doing test upload to %s", url)
+            resp1 = await self.file_client.put(
+                url,
+                content=b"",
+                headers=headers,
+                follow_redirects=False,
+            )
+            if resp1.status_code == 307 and resp1.headers["Location"]:
+                log.info("got file upload redirect from api")
+                url = resp1.headers["Location"]
+
+        log.info("doing real upload to %s", url)
+        # set connect timeout to slightly more than a multiple of 3 to avoid
+        # aligning perfectly with TCP retransmission timer
+        timeout = httpx.Timeout(10.0, read=15.0)
+        resp = await self.file_client.put(
+            url,
+            content=ChunkFileReader(fh),
+            headers=headers,
+            timeout=timeout,
+        )
+        # TODO: if file size is >1MB, show upload throughput
+        resp.raise_for_status()
+
+        # Try to extract the final asset URL from the `Location` header
+        # otherwise fallback to the URL of the final request.
+        final_url = str(resp.url)
+        if "location" in resp.headers:
+            final_url = resp.headers.get("location")
+
+        # strip any signing gubbins from the URL
+        return urlparse(final_url)._replace(query="").geturl()
+
+    # this previously lived in json.upload_files, but it's clearer here
+    # this is a great pattern that should be adopted for input files
```

Comment:
What pattern is being referred to here?

Reply:
recursively iterating through the entire object rather than only the top level